server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java

```diff
@@ -0,0 +1,386 @@
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.transport.TransportAddress;
+import org.opensearch.discovery.PluginRequest;
+import org.opensearch.discovery.PluginResponse;
+import org.opensearch.extensions.ExtensionsSettings.Extension;
+import org.opensearch.index.IndexModule;
+import org.opensearch.index.IndexService;
+import org.opensearch.index.IndicesModuleNameResponse;
+import org.opensearch.index.IndicesModuleRequest;
+import org.opensearch.index.IndicesModuleResponse;
+import org.opensearch.index.shard.IndexEventListener;
+import org.opensearch.indices.cluster.IndicesClusterStateService;
+import org.opensearch.node.ReportingService;
+import org.opensearch.plugins.PluginInfo;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+
+/**
+ * The main class for Plugin Extensibility
+ *
+ * @opensearch.internal
+ */
+public class ExtensionsOrchestrator implements ReportingService<PluginsAndModules> {
+    public static final String REQUEST_EXTENSION_ACTION_NAME = "internal:discovery/extensions";
+    public static final String INDICES_EXTENSION_POINT_ACTION_NAME = "indices:internal/extensions";
+    public static final String INDICES_EXTENSION_NAME_ACTION_NAME = "indices:internal/name";
+    public static final String REQUEST_EXTENSION_CLUSTER_STATE = "internal:discovery/clusterstate";
+    public static final String REQUEST_EXTENSION_LOCAL_NODE = "internal:discovery/localnode";
+    public static final String REQUEST_EXTENSION_CLUSTER_SETTINGS = "internal:discovery/clustersettings";
+
+    private static final Logger logger = LogManager.getLogger(ExtensionsOrchestrator.class);
+
+    /**
+     * Enum for Extension Requests
+     *
+     * @opensearch.internal
+     */
+    public static enum RequestType {
+        REQUEST_EXTENSION_CLUSTER_STATE,
+        REQUEST_EXTENSION_LOCAL_NODE,
+        REQUEST_EXTENSION_CLUSTER_SETTINGS,
+        CREATE_COMPONENT,
+        ON_INDEX_MODULE,
+        GET_SETTINGS
+    };
+
+    private final Path extensionsPath;
+    final List<DiscoveryExtension> extensionsList;
+    List<DiscoveryExtension> extensionsInitializedList;
+    TransportService transportService;
+    ClusterService clusterService;
+
+    public ExtensionsOrchestrator(Settings settings, Path extensionsPath) throws IOException {
+        logger.info("ExtensionsOrchestrator initialized");
+        this.extensionsPath = extensionsPath;
+        this.transportService = null;
+        this.extensionsList = new ArrayList<DiscoveryExtension>();
+        this.extensionsInitializedList = new ArrayList<DiscoveryExtension>();
```

Comment:
I think we end up using initialized extensions by default, so I would flip this to `uninitializedExtensions` and `extensions`.